generator/internal/api/dependencies.go

```diff
@@ -0,0 +1,164 @@
+//
+// We can think of the `model` as a directed graph, with nodes for each of the
+// model elements. Finding required elements is a graph traversal problem.
+//
+// First observe that:
+//   - if we are given a service, we want to include all of its methods.
+//   - if we are given a method, we need to include its service, but not
+//     necessarily any of its sibling methods.
+//
+// This implies that simply fanning out over the nodes is not sufficient.
+//
+// We resolve this by making two passes. In the first pass, we fan out over all
+// child elements of the given elements. In the second pass, we add all required
+// parents of any found elements.
+//
+// In the first pass, we compute the [reachable set] of the given `ids`, with
+// edges from...
+//
+// - a service to each of its methods
+// - a method to its request/response messages
+// - an LRO method to its metadata/response type messages
+// - a message to each of its fields' types (if they are a message or enum)
+// - a parent message to each of its nested messages
+// - a parent message to each of its nested enums
+//
+// In the second pass, we compute the reachable set of the nodes found from the
+// first pass. This time, the graph has edges from...
+//
+// - a method to its service
+// - a child message to its parent message
+// - a child enum to its parent message
+// - a message to each of its fields' types (if they are a message or enum)
+//   - this edge is only necessary because we are too lazy to support pruning
+//     the fields of a message.
+//
+// [reachable set]: https://en.wikipedia.org/wiki/Reachability
+func FindDependencies(model *API, ids []string) (map[string]bool, error) {
+	includedIDs := map[string]bool{}
+	candidates := []string{}
+
+	add := func(id string) {
+		if _, ok := includedIDs[id]; !ok {
+			candidates = append(candidates, id)
+		}
+		includedIDs[id] = true
+	}
+
+	// Seed with the given ids
+	for _, id := range ids {
+		add(id)
+	}
+
+	// Fan out over all child elements of the given nodes.
+	for len(candidates) > 0 {
+		id := candidates[len(candidates)-1]
+		candidates = candidates[0 : len(candidates)-1]
+
+		// Recurse one level, depending on the input type.
+		service, ok := model.State.ServiceByID[id]
+		if ok {
+			for _, method := range service.Methods {
+				add(method.ID)
+			}
+			continue
+		}
+
+		method, ok := model.State.MethodByID[id]
+		if ok {
+			add(method.InputTypeID)
+			add(method.OutputTypeID)
+			if method.OperationInfo != nil {
+				add(method.OperationInfo.MetadataTypeID)
+				add(method.OperationInfo.ResponseTypeID)
+			}
+			continue
+		}
+
+		message, ok := model.State.MessageByID[id]
+		if ok {
+			for _, msg := range message.Messages {
+				add(msg.ID)
+			}
+			for _, enum := range message.Enums {
+				add(enum.ID)
+			}
+			for _, field := range message.Fields {
+				if field.Typez == ENUM_TYPE || field.Typez == MESSAGE_TYPE {
+					add(field.TypezID)
+				}
+			}
```

Comment:
Hmm, actually, I think this should only be in the second pass.

Reply:
Considering: 

```proto
message A {
  B b = 1;
}

message B {
  enum C {}
}
```

Given `A`, should we generate `B.C` ?

Ummm... I think so. Which means we want this code.

Reply:
> I suspect we will largely use this for methods and services anyway.

Yep.

> a bit subtle

Yeah... I like the formulation of: `the minimum set that satisfies "I can use the inputs"`.

But when I think about what is the minimum set that satisfies I can use a message, it is the message with no fields. The minimum set that satisfies I can use a service is the service with no methods.

So I think my formulation is more like `the minimum set that satisfies "I can use the inputs completely"`. :shrug: 

And, even in that formulation, we do not need `B` to imply `B.C`... So I think I will drop that edge.

> Did I convince you that `B` should imply `B.C` ?

I am the one who has been convinced.